### PR TITLE
[corlib] Type.GetType parser and resolver improvements

### PR DESCRIPTION
**Pull request by lambdageek**
We have two type parsers the one invoked by `Type.GetType(string)` (and some others) which is implemented in C in `reflection.c`, and the one invoked by `Type.GetType(string, Func<...>, Func<...>)` implemented in managed code as `System.TypeSpec`.

In both cases `string→Type` conversion goes in two stages: parsing and resolution.  Parsing just changes the string into an AST.  And resolution crawls the AST and constructs the appropriate types.

This fixes [Bugzilla #40060](https://bugzilla.xamarin.com/show_bug.cgi?id=40060)

1. Fix the resolver on the native side for types like `"System.Generic.Collections.Dictionary[System.Int32,MyType]"` where the generic type is from corlib and a type argument is from the current assembly.
2. Fix `get_caller_no_reflection` to work correctly when the call stack has a System.Reflection method as the immediate caller of the icall.
3. Fix the resolver on the native side to work when it's called from corlib's System namespace, not just System.Reflection.
4. Fix the managed parser to work with types like `"MyGeneric[MyType]"` where the generic argument is not an assembly qualified name or there's a mixture of aqn and non-aqn type arguments.  Also rationalize how the code handles assembly qualification in general.
5. Test cases

### CODE REVIEW

File: mono/metadata/icall.c [hunk 1]
```diff
@@ -1403,10 +1452,12 @@ ves_icall_System_Type_internal_from_name (MonoString *name,
 	/* mono_reflection_parse_type() mangles the string */
 	if (!parsedOk) {
 		mono_reflection_free_type_info (&info);
-		g_free (str);
 		if (throwOnError) {
-			mono_set_pending_exception (mono_get_exception_argument("typeName", "failed parse"));
+			mono_error_init (&error);
```
Comment (kumpera) [inline mono/metadata/icall.c]: Is this necessary?

**Comment by kumpera** [comment]
LGTM

File: mono/metadata/icall.c (hunk 1 above)
Comment (lambdageek) [inline mono/metadata/icall.c]: No, not really.  The main thing I was after was to include `str` in the `"failed parse: %s"` error message.  This seemed a bit less annoying than `g_strdup_printf()/free()`.

Comment (lambdageek) [inline mono/metadata/icall.c]: Oh if you mean do I need to do `mono_error_init`, then the answer is yes: nothing else initializes the `flags` field.

**Comment by lambdageek** [comment]
Thanks!

@monojenkins merge

**Pull request merged**